pommerman/network/ion_client/network.py

```diff
@@ -0,0 +1,186 @@
+        self.mode = decoded_msg["mode"]
+
+    def waitm(self):
+        """Description: Wait for a response from the server regarding a match 
+        request"""
+        self.lock.acquire()
+        try:
+            r = rapidjson.loads(self.ws_.recv())
+            if r["intent"] == constants.NetworkCommands.match_start.value:
+                self.match_id = r["match_id"]
+        except:
+            self.lock.release()
+            raise Exception(constants.Exceptions.net_server_closed.value)
+        self.lock.release()
+
+    def match_get(self):
+        """Description: Get the next step of the match"""
+        self.lock.acquire()
+        try:
+            message_recieved = self.ws_.recv()
+        except:
+            self.lock.release()
+            raise Exception(constants.Exceptions.net_respond_fail.value)
+        self.lock.release()
+        try:
+            r = rapidjson.loads(str(gzip.decompress(message_recieved), "utf-8"))
+        except:
+            try:
+                message_recieved = rapidjson.loads(message_recieved)
+                if message_recieved[
+                    "intent"
+                ] == constants.NetworkCommands.match_end.value:
+                    return [
+                        2,
+                        int(message_recieved["reward"]),
+                        int(message_recieved["agent"])
+                    ]
+            except:
+                raise Exception(constants.Exceptions.net_invalid_response.value)
+        if r["d"]:
+            return [1]
+        obs = r["o"]
+        obs["teammate"] = pommerman.constants.Item[obs["teammate"]]
+        # Note: If position is not tuple SimpleAgent *will* error out
+        obs["position"] = tuple(obs["position"])
+        for x, y in enumerate(obs["enemies"]):
+            obs["enemies"][x] = pommerman.constants.Item[y]
+        for i in ["board", "bomb_life", "bomb_blast_strength"]:
+            obs[i] = numpy.asarray(obs[i])
+        return [0, obs, r["i"]]  # r["i"] = Turn ID
+
+    def send_move(self, action, turn_id):
+        """Argument:  
+        * action: The action that has to be sent  
+        * turn_id: The ID of the step taken by the server (To sync up the \
+action of the agent and server)  
+        Description: Send the action to the server for playing out"""
+        self.lock.acquire()
+        try:
+            self.ws_.send(
+                rapidjson.dumps(
+                    {
+                        "intent": constants.NetworkCommands.match_act.value,
+                        "player_id": self.id,
+                        "act": action,
+                        "match_id": self.match_id,
+                        "turn_id": turn_id
+                    }
+                )
+            )
+        except:
+            self.lock.release()
+            if not self.ws_.connected:
+                raise Exception(constants.Exceptions.net_server_closed.value)
+        self.lock.release()
```

Comment:
Same re Exception and Finally questions.

Reply:
^^^^^^^^^^^^